core/src/ctx/context.rs

```diff
@@ -246,6 +251,19 @@ impl<'a> Context<'a> {
 		self.iteration_stage = Some(is);
 	}
 
+	pub(crate) fn set_transaction_mut(&mut self, txn: Transaction) {
+		self.transaction = Some(txn);
+	}
+
+	pub(crate) fn set_transaction(mut self, txn: Transaction) -> Self {
+		self.transaction = Some(txn);
+		self
+	}
```

Comment:
We have two setter functions to provide the most appropriate behaviour depending on the context.